Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -928,7 +928,7 @@ Vector<CSS::BackgroundLayerData> const* Document::background_layers() const
 
 void Document::update_base_element(Badge<HTML::HTMLBaseElement>)
 {
-    JS::GCPtr<HTML::HTMLBaseElement const> base_element;
+    HTML::HTMLBaseElement const* base_element;
```

Comment:
Three problems here:

1. We should not allow capturing raw pointers by reference either! They have the exact same problem as GCPtr, which is: if the lambda closure outlives the referee, the capture is unsafe!

2. The capture-by-reference here is intentional, since we want to overwrite the local from inside the lambda. The reason it is safe is because we know the lambda will be invoked synchronously and cannot outlive the `base_element` local. I think we would need some kind of annotation to inform LibJSGCVerifier that this is okay. Perhaps we could invent a `[[clang::sync_invoked]]` attribute and slap it on the callback type? Then we could say that the warning is only relevant when the lambda doesn't have this annotation. (This could potentially allow us to catch a whole new class of bugs in our codebase actually! Since the problem is not specific to GC, but a frequently occurring general C++ problem.)

3. Switching to a raw pointer actually introduces a bug: it's uninitialized. As `GCPtr`, it was null by default. :sweat_smile:

Reply:
Hah, didn't even think about the initialization. I'm not sure if this is true, but it feels like most callbacks are called synchronously, so maybe inverting it to `[[clang::deferred_invoke]]` would make more sense?

Reply:
> Hah, didn't even think about the initialization. I'm not sure if this is true, but it feels like most callbacks are called synchronously, so maybe inverting it to `[[clang::deferred_invoke]]` would make more sense?

Hmm.. you might be right that sync invocation is more common but on the other hand, a "paranoid by default" approach would probably catch more bugs. :)

Reply:
And yeah, this does indeed expand the scope (and usefulness!) of `LibJSGCVerifier`. Perhaps we could rename it to something more general, e.g `BugCatcher` or similar

Reply:
With the new commits, all of the "fixes" I had before no longer apply